cluster/state.go

```diff
@@ -249,3 +134,106 @@ func GetK8sVersion(localConfigPath string, k8sWrapTransport k8s.WrapTransport) (
+		if err := pki.GenerateKubeAPICertificate(ctx, pkiCertBundle, *rkeConfig, "", ""); err != nil {
+			return nil, err
+		}
+		// Regenerating kubeadmin certificates/config
+		if err := pki.GenerateKubeAdminCertificate(ctx, pkiCertBundle, *rkeConfig, configPath, configDir); err != nil {
+			return nil, err
+		}
+		newState.DesiredState.CertificatesBundle = pkiCertBundle
+	}
+	newState.CurrentState = oldState.CurrentState
+	return newState, nil
+}
+
+func (s *RKEFullState) WriteStateFile(ctx context.Context, statePath string) error {
+	stateFile, err := json.MarshalIndent(s, "", "  ")
+	if err != nil {
+		return fmt.Errorf("Failed to Marshal state object: %v", err)
+	}
+	logrus.Debugf("Writing state file: %s", stateFile)
+	if err := ioutil.WriteFile(statePath, stateFile, 0640); err != nil {
+		return fmt.Errorf("Failed to write state file: %v", err)
+	}
+	log.Infof(ctx, "Successfully Deployed state file at [%s]", statePath)
+	return nil
+}
+
+func GetStateFilePath(configPath, configDir string) string {
+	if configPath == "" {
+		configPath = pki.ClusterConfig
+	}
+	baseDir := filepath.Dir(configPath)
+	if len(configDir) > 0 {
+		baseDir = filepath.Dir(configDir)
+	}
+	fileName := filepath.Base(configPath)
+	baseDir += "/"
+	fullPath := fmt.Sprintf("%s%s", baseDir, fileName)
+	trimmedName := strings.TrimSuffix(fullPath, filepath.Ext(fullPath))
+	return trimmedName + stateFileExt
+}
+
+func ReadStateFile(ctx context.Context, statePath string) (*RKEFullState, error) {
+	rkeFullState := &RKEFullState{}
+	fp, err := filepath.Abs(statePath)
+	if err != nil {
+		return rkeFullState, fmt.Errorf("failed to lookup current directory name: %v", err)
+	}
+	file, err := os.Open(fp)
+	if err != nil {
+		return rkeFullState, fmt.Errorf("Can not find RKE state file: %v", err)
+	}
+	defer file.Close()
+	buf, err := ioutil.ReadAll(file)
+	if err != nil {
+		return rkeFullState, fmt.Errorf("failed to read state file: %v", err)
+	}
+	if err := json.Unmarshal(buf, rkeFullState); err != nil {
+		return rkeFullState, fmt.Errorf("failed to unmarshal the state file: %v", err)
+	}
+	rkeFullState.DesiredState.CertificatesBundle = pki.TransformPEMToObject(rkeFullState.DesiredState.CertificatesBundle)
+	rkeFullState.CurrentState.CertificatesBundle = pki.TransformPEMToObject(rkeFullState.CurrentState.CertificatesBundle)
+	return rkeFullState, nil
+}
+
+func RemoveStateFile(ctx context.Context, statePath string) {
+	log.Infof(ctx, "Removing state file: %s", statePath)
+	if err := os.Remove(statePath); err != nil {
+		logrus.Warningf("Failed to remove state file: %v", err)
+		return
+	}
+	log.Infof(ctx, "State file removed successfully")
+}
+
+func RemoveLegacyStateFromKubernets(ctx context.Context, kubeConfigPath string, k8sWrapTransport k8s.WrapTransport) error {
+	k8sClient, err := k8s.NewClient(kubeConfigPath, k8sWrapTransport)
+	if err != nil {
+		return fmt.Errorf("Failed to create Kubernetes Client: %v", err)
+	}
+	return k8s.DeleteConfigMap(k8sClient, StateConfigMapName)
```

Comment:
what are the implications of leaving the legacy state around? Thinking of situations when user upgrades rancher, triggers rke cluster update, and then rollbacks rancher to the older version where the state was expected to exist on the cluster side.

Reply:
that's true, i will keep the config map even after upgrade